diff --git a/src/libretro/video.c b/src/libretro/video.c
index bb42a8e..c0f6755 100644
--- a/src/libretro/video.c
+++ b/src/libretro/video.c
@@ -366,23 +366,13 @@ static void select_display_mode(int width,int height,int depth,int attributes,in
 		gfx_height=height;
 	}
 
-	/* vector games use 640x480 as default */
+	/* vector games use 320x240 as default */
 	if (vector_game && !iOS_fixedRes)
 	{
-		if(safe_render_path)
-		{
-		   gfx_width = 640;
-		   gfx_height = 480;
-		   emulated_width = 640;
-		   emulated_height = 480;
-		}
-		else
-		{
-		   gfx_width = 320;
-		   gfx_height = 240;
-		   emulated_width = 320;
-		   emulated_height = 240;
-		}
+		gfx_width = 320;
+		gfx_height = 240;
+		emulated_width = 320;
+		emulated_height = 240;
 	}
 
 	gp2x_set_video_mode(16,gfx_width,gfx_height);
diff --git a/src/mame.c b/src/mame.c
index 4d0cc3c..333a457 100644
--- a/src/mame.c
+++ b/src/mame.c
@@ -64,8 +64,8 @@ int run_game(int game)
 
     if(!iOS_fixedRes)
     {
-	  options.vector_width = 640;
-	  options.vector_height = 480;
+	  options.vector_width = 320;
+	  options.vector_height = 240;
     }
     else
     {
