diff --git a/platform/libretro/Makefile b/platform/libretro/Makefile
index 40805b8..d890e0c 100644
--- a/platform/libretro/Makefile
+++ b/platform/libretro/Makefile
@@ -123,7 +123,36 @@ else ifeq ($(platform), miyoomini)
    CXXFLAGS += -marm -mtune=cortex-a7 -mfpu=neon-vfpv4 -mfloat-abi=hard -march=armv7ve -D_NEED_FULL_PATH_
    fpic := -fPIC
    SHARED := -shared -Wl,--version-script=link.T
-
+else ifneq (,$(findstring armv,$(platform)))
+	TARGET := $(TARGET_NAME)_libretro.so
+	SHARED := -shared -Wl,--no-undefined,-Bsymbolic
+	fpic := -fPIC
+	ifneq (,$(findstring cortexa5,$(platform)))
+		CFLAGS += -marm -mcpu=cortex-a5
+		ASFLAGS += -mcpu=cortex-a5
+	else ifneq (,$(findstring cortexa8,$(platform)))
+		CFLAGS += -marm -mcpu=cortex-a8
+		ASFLAGS += -mcpu=cortex-a8
+	else ifneq (,$(findstring cortexa9,$(platform)))
+		CFLAGS += -marm -mcpu=cortex-a9
+		ASFLAGS += -mcpu=cortex-a9
+	else ifneq (,$(findstring cortexa15a7,$(platform)))
+		CFLAGS += -marm -mcpu=cortex-a15.cortex-a7
+		ASFLAGS += -mcpu=cortex-a15.cortex-a7
+	else
+		CFLAGS += -marm
+	endif
+	ifneq (,$(findstring neon,$(platform)))
+		CFLAGS += -mfpu=neon
+		ASFLAGS += -mfpu=neon
+	endif
+	ifneq (,$(findstring softfloat,$(platform)))
+		CFLAGS += -mfloat-abi=softfp
+		ASFLAGS += -mfloat-abi=softfp
+	else ifneq (,$(findstring hardfloat,$(platform)))
+		CFLAGS += -mfloat-abi=hard
+		ASFLAGS += -mfloat-abi=hard
+	endif
 else ifeq ($(platform), gcw0)
    TARGET := $(TARGET_NAME)_libretro_gcw0.so
    CC = /opt/gcw0-toolchain/usr/bin/mipsel-linux-gcc
