diff --git a/platform/libretro/Makefile b/platform/libretro/Makefile
index 420f9b7..b247ec0 100644
--- a/platform/libretro/Makefile
+++ b/platform/libretro/Makefile
@@ -123,7 +123,11 @@ else ifeq ($(platform), miyoomini)
    CXXFLAGS += -marm -mtune=cortex-a7 -mfpu=neon-vfpv4 -mfloat-abi=hard -march=armv7ve
    fpic := -fPIC
    SHARED := -shared -Wl,--version-script=link.T
-
+else ifeq ($(platform), classic_armv7_a7)
+   TARGET := $(TARGET_NAME)_libretro.so
+   CXXFLAGS += -marm -mtune=cortex-a7 -mfpu=neon-vfpv4 -mfloat-abi=hard -march=armv7ve
+   fpic := -fPIC
+   SHARED := -shared -Wl,--version-script=link.T
 else ifeq ($(platform), gcw0)
    TARGET := $(TARGET_NAME)_libretro_gcw0.so
    CC = /opt/gcw0-toolchain/usr/bin/mipsel-linux-gcc
