diff --git a/in_sdl.c b/in_sdl.c
index a84c781..22d3b01 100644
--- a/in_sdl.c
+++ b/in_sdl.c
@@ -476,10 +476,8 @@ static int in_sdl_menu_translate(void *drv_data, int keycode, char *charcode)
 	else
 	{
 		for (i = 0; i < map_len; i++) {
-			if (map[i].key == keycode) {
-				ret = map[i].pbtn;
-				break;
-			}
+			if (map[i].key == keycode)
+				return map[i].pbtn;
 		}
 
 		if (charcode != NULL && (unsigned int)keycode < SDLK_LAST &&
